File: BEmuJava/bemu/src/main/java/com/bloomberglp/blpapi/MarketMessageSubscriptionCanceled.java
//------------------------------------------------------------------------------
// <copyright project="BEmu_maven" file="/BEmu_maven/bemu/src/main/java/com/bloomberglp/blpapi/MarketMessageSubscriptionCanceled.java" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

package com.bloomberglp.blpapi;

import com.bloomberglp.blpapi.Message;
import com.bloomberglp.blpapi.Subscription;
import com.bloomberglp.blpapi.Name;
import com.bloomberglp.blpapi.Element;

public class MarketMessageSubscriptionCanceled extends Message
{
    private final String _topicName;
    private final MarketElementSubscriptionCancelReason _reason;

    MarketMessageSubscriptionCanceled(Subscription sub) throws Exception
    {
    	super(new Name("SubscriptionTerminated"), sub.correlationID(), null);
    	
        this._topicName = sub.security();
        this._reason = new MarketElementSubscriptionCancelReason();
    }
    
	public int numElements()
	{
		return 1;
	}
    
	public String topicName()
	{
		return this._topicName;
	}
	
	public boolean hasElement(String name, boolean excludeNullElements) throws Exception
	{
		return this._reason.name().toString().equals(name);
	}
	
	public Element getElement(String name) throws Exception
	{
		if(name.equals(this._reason.name().toString()))
			return this._reason;
			
		else
			return super.getElement(name);
	}
	
    public String toString()
    {
        StringBuilder result = new StringBuilder();
        result.append(String.format("%s = {%s", this.messageType().toString(), System.getProperty("line.separator")));
        try
        {
        	result.append(this._reason.prettyPrint(1));
		}
	    catch (Exception e) { }
        result.append("}" + System.getProperty("line.separator"));

        return result.toString();
    }

}

File: BEmu_cpp/src/MarketDataRequest/MarketMessageSubscriptionCanceled.cpp
//------------------------------------------------------------------------------
// <copyright project="BEmu_cpp" file="src/MarketDataRequest/MarketMessageSubscriptionCanceled.cpp" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

#include "MarketDataRequest/MarketMessageSubscriptionCanceled.h"
#include "MarketDataRequest/MarketElementSubscriptionCancelReason.h"
#include "BloombergTypes/Name.h"
#include "BloombergTypes/Subscription.h"

namespace BEmu
{
	namespace MarketDataRequest
	{
		MarketMessageSubscriptionCanceled::MarketMessageSubscriptionCanceled(const Subscription& sub) :
			MessagePtr(Name("SubscriptionTerminated"), sub.correlationId()), CanConvertToStringType(sub.security()),
			_reason(new MarketElementSubscriptionCancelReason())
		{
			this->_topicName = sub.security();
		}

		MarketMessageSubscriptionCanceled::~MarketMessageSubscriptionCanceled()
		{
		}

		size_t MarketMessageSubscriptionCanceled::numElements() const
		{
			return 1;
		}

		const char* MarketMessageSubscriptionCanceled::topicName() const
		{
			return this->ValueAsString().c_str();
		}

		boost::shared_ptr<ElementPtr> MarketMessageSubscriptionCanceled::asElement() const
		{
			throw messageEx;
		}

		std::ostream& MarketMessageSubscriptionCanceled::print(std::ostream& stream, int level, int spacesPerLevel) const
		{
			stream << "SubscriptionTerminated = {"  << std::endl;
			this->_reason->print(stream, level + 1, spacesPerLevel);
			stream << "}" << std::endl;

			return stream;
		}

	}
}